pkg/chaosdaemon/tasks/process_group_handler.go

```diff
@@ -0,0 +1,135 @@
+
+func (s SysPID) ToID() string {
+	return strconv.FormatUint(uint64(s), 10)
+}
+
+// ChaosOnProcessGroup is used for inject a chaos on a linux process group.
+// Fork is used for create a new chaos on child process.
+// Assign is used for update a chaos on child process.
+type ChaosOnProcessGroup interface {
+	Fork() (ChaosOnProcessGroup, error)
+	Assign
+
+	Injectable
+	Recoverable
+}
+
+// ProcessGroupHandler implements injecting & recovering on a linux process group.
+type ProcessGroupHandler struct {
+	LeaderProcess ChaosOnProcessGroup
+	childMap      map[IsID]ChaosOnProcessGroup
+	Logger        logr.Logger
+}
+
+func NewProcessGroupHandler(logger logr.Logger, leader ChaosOnProcessGroup) ProcessGroupHandler {
+	return ProcessGroupHandler{
+		LeaderProcess: leader,
+		childMap:      make(map[IsID]ChaosOnProcessGroup),
+		Logger:        logr.New(logger.GetSink()),
+	}
+}
+
+// Inject try to inject the main process and then try to inject child process.
+// If something wrong in injecting a child process, Inject will just log error & continue.
+func (gp *ProcessGroupHandler) Inject(pid IsID) error {
+	sysPID, ok := pid.(SysPID)
+	if !ok {
+		return ErrNotTypeSysID.WrapInput(pid).Err()
+	}
+
+	err := gp.LeaderProcess.Inject(sysPID)
+	if err != nil {
+		return cerr.FromErr(err).Wrapf("inject main process: %v", sysPID).Err()
+	}
+
+	childPIDs, err := util.GetChildProcesses(uint32(sysPID), gp.Logger)
+	if err != nil {
+		return cerr.NotFound("child process").WrapErr(err).Err()
+	}
+
+	for _, childPID := range childPIDs {
+		childSysPID := SysPID(childPID)
+		if childProcessChaos, ok := gp.childMap[childSysPID]; ok {
+			err := gp.LeaderProcess.Assign(childProcessChaos)
+			if err != nil {
+				gp.Logger.Error(err, "failed to assign old child process")
+				continue
+			}
+			err = childProcessChaos.Inject(childSysPID)
+			if err != nil {
+				gp.Logger.Error(err, "failed to inject old child process")
+			}
+		} else {
+			childProcessChaos, err := gp.LeaderProcess.Fork()
+			if err != nil {
+				gp.Logger.Error(err, "failed to create child process")
+				continue
+			}
+			err = childProcessChaos.Inject(pid)
+			if err != nil {
+				gp.Logger.Error(err, "failed to inject new child process")
+				continue
+			}
+			gp.childMap[childSysPID] = childProcessChaos
+		}
+	}
+	return nil
+}
+
+// Recover try to recover the main process and then try to recover child process.
+func (gp *ProcessGroupHandler) Recover(pid IsID) error {
+	sysPID, ok := pid.(SysPID)
+	if !ok {
+		return ErrNotTypeSysID.WrapInput(pid).Err()
+	}
+	err := gp.LeaderProcess.Recover(pid)
+	if err != nil {
+		return cerr.FromErr(err).Wrapf("recovery main process : %v", pid).Err()
+	}
+
+	childPids, err := util.GetChildProcesses(uint32(sysPID), gp.Logger)
+	if err != nil {
+		return cerr.NotFound("child process").WrapErr(err).Err()
+	}
+
+	for _, childPID := range childPids {
+		childSysPID := SysPID(childPID)
+		if childProcessChaos, ok := gp.childMap[childSysPID]; ok {
```

Comment:
Maybe we could iterate through the `childMap` directly (without finding child processes first), it should be able to handle exited processes.

Reply:
I agree with you.

Reply:
Fixed.